File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/22-UnionFind/LeetCode/LeetCode-1971.py
# Implement a Union/Find DS
# Find set with path compression
class DS:
    def __init__(self, n):
        self.sets = [-1]*n

    def find(self, u):
        if (self.sets[u] < 0): return u
        
        s = self.find(self.sets[u])
        self.sets[u] = s
        return s
    
    def join(self, u, v):
        if (self.sets[u] < self.sets[v]):
            self.sets[u] += self.sets[v]
            self.sets[v] = u
        else:
            self.sets[v] += self.sets[u]
            self.sets[u] = v

    # Assumes that "u" is a valid set head
    def size(self, u):
        return -self.sets[u]

    
class Solution:
    def validPath(self, n: int, edges: List[List[int]], start: int, end: int) -> bool:
        """
        For a path to exist between start & end, they have to be
        on the same connected component, i.e., same set
        """
        ds = DS(n)
        
        for u, v in edges:
            set1 = ds.find(u)
            set2 = ds.find(v)
            if set1 != set2:
                ds.join(set1, set2)
        return ds.find(start) == ds.find(end)
        

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/22-UnionFind/LeetCode/LeetCode-1971.cpp
/// Disjoint set class
class DS {
    vector<int> sets;
public:
    DS(int n){sets.resize(n, -1);}
    
    // Find set with path compression
    int find(int u){
        if (sets[u] < 0) return u;
        
        int s = find(sets[u]);
        sets[u] = s;
        return s;
    } // end-findSet
    
    void join(int u, int v){
        if (u == v) return;
        if (sets[u] < sets[v]){
            sets[u] += sets[v];
            sets[v] = u;
        } else {
            sets[v] += sets[u];
            sets[u] = v;
        } //end-else
    } // end-join
    
    // Assume that "u" is a valid set head
    int size(int u){return -sets[u];}
}; // end-DS

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int start, int end) {
        /// For a path to exist between start & end, they have to be
        /// on the same connected component, i.e., same set
        DS ds(n);

        for (int i=0; i<edges.size(); i++){
            int set1 = ds.find(edges[i][0]);
            int set2 = ds.find(edges[i][1]);
            ds.join(set1, set2);
        } // end-for
        
        return ds.find(start) == ds.find(end);
    } // end-validPath
};